lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp

```diff
@@ -243,6 +384,22 @@ struct PrintOpConversion
     SmallVector<Value> operands{prefixString, bufferPtr};
     call(funcOp, operands);
   }
+
+private:
+  Value getOrAddFormatStrToModule(Location loc,
+                                  ConversionPatternRewriter &rewriter,
+                                  StringRef str) const {
+    auto [it, inserted] = formatStrCache.insert({str, Value{}});
+    if (inserted) {
+      it->second = LLVM::addStringToModule(loc, rewriter, "printfFormat_", str);
+    }
+    return it->second;
+  }
+
+  // Cache the llvm Value for each format string to avoid making duplicates in
+  // the module.  Sorry about `mutable`, but the top-level matchAndRewrite
+  // function is const, so it virally comes down here.
+  mutable llvm::StringMap<Value> formatStrCache;
```

Comment:
Caching will break if the new print function is not dominated by the value that has been cached. For instance it will always fail if we are printing from different functions. In general doing caching during rewrite patterns is not trivial, we could cache only the global value but this is probably a premature  optimization.

Reply:
addStringToModule adds the string as a global, so shouldn't this work?  https://github.com/openai/triton/blob/020f43d5a37c2f1cce5c70fcda17ec57b000755c/lib/Conversion/TritonGPUToLLVM/Utility.cpp#L354

(The reason I wanted to cache this is otherwise we get BLOCK_SIZE copies of the format-string.  Or otherwise we have to rewrite it so we generate the format-string only once per block, but that's also bug-prone and harder to read.)

Reply:
but it creates operations that are within a block:
```
Value stringStart =
      rewriter.create<LLVM::GEPOp>(UnknownLoc::get(ctx), ptr_ty(i8_ty),
                                   globalPtr, SmallVector<Value>({zero, zero}));
  return stringStart;
```
if those are used by an instruction that isn't dominated by the cached ops it would break SSA. Can you change the code so that this gets called outside the loop?

Reply:
Indeed you're completely correct.  Done, thanks.